File: sdks/node/tests/producer/producer.test.ts
import {
  Producer,
  recordValidationMetrics,
  recordRejection,
  getMetrics,
  resetMetrics,
} from "../../src/producer/producer";
import {
  shouldValidatePath,
  matchesPathFilter,
  type Validator,
  type ProducerValidationResult,
  type Interaction,
} from "../../src/producer/types";

/**
 * Mock validator that always returns valid.
 */
function createMockValidator(
  validateFn?: (
    schemaId: string,
    interaction: Interaction,
  ) => Promise<ProducerValidationResult>,
): Validator {
  return {
    validate:
      validateFn ||
      (async () => ({
        valid: true,
        errors: [],
      })),
  };
}

/**
 * Mock validator that always returns invalid.
 */
function createInvalidValidator(
  errors: string[] = ["Validation error"],
): Validator {
  return {
    validate: async () => ({
      valid: false,
      errors,
    }),
  };
}

/**
 * Mock Express-style response object.
 */
function createMockResponse() {
  const res: any = {
    statusCode: 200,
    headersSent: false,
    headers: {} as Record<string, string>,
    body: null as any,
    status(code: number) {
      res.statusCode = code;
      return res;
    },
    json(data: any) {
      res.body = data;
      res.headersSent = true;
      return res;
    },
    setHeader(name: string, value: string) {
      res.headers[name] = value;
    },
    end(data?: any) {
      if (data) res.body = data;
      res.headersSent = true;
    },
  };
  return res;
}

/**
 * Mock Express-style request object.
 */
function createMockRequest(method: string, path: string, body?: any) {
  return {
    method,
    url: path,
    path,
    headers: { "content-type": "application/json" },
    body,
  };
}

describe("Producer", () => {
  describe("constructor", () => {
    it("should create producer with valid config", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
      });
      expect(producer).toBeDefined();
      expect(producer.mode).toBe("strict");
    });

    it("should throw error when schemaId is missing", () => {
      expect(() => {
        new Producer({
          schemaId: "",
          validator: createMockValidator(),
        });
      }).toThrow("schemaId is required");
    });

    it("should throw error when validator is missing", () => {
      expect(() => {
        new Producer({
          schemaId: "test-schema",
          validator: undefined as any,
        });
      }).toThrow("validator is required");
    });

    it("should apply default config values", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
      });
      expect(producer.mode).toBe("strict");
    });

    it("should use custom mode when provided", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "warn",
      });
      expect(producer.mode).toBe("warn");
    });
  });

  describe("validateRequest", () => {
    it("should validate a valid request", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
      });

      const result = await producer.validateRequest(
        "POST",
        "/users",
        { "content-type": "application/json" },
        { name: "test" },
      );

      expect(result.valid).toBe(true);
      expect(result.type).toBe("request");
    });

    it("should return invalid for invalid request", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createInvalidValidator(["missing required field: email"]),
      });

      const result = await producer.validateRequest(
        "POST",
        "/users",
        { "content-type": "application/json" },
        { name: "test" },
      );

      expect(result.valid).toBe(false);
      expect(result.errors).toContain("missing required field: email");
      expect(result.type).toBe("request");
    });

    it("should skip validation when validateRequest is false", async () => {
      const validateFn = jest.fn();
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(validateFn),
        validateRequest: false,
      });

      const result = await producer.validateRequest("POST", "/users", {}, {});

      expect(result.valid).toBe(true);
      expect(validateFn).not.toHaveBeenCalled();
    });

    it("should pass correct interaction to validator", async () => {
      let capturedInteraction: Interaction | null = null;
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(async (schemaId, interaction) => {
          capturedInteraction = interaction;
          return { valid: true };
        }),
      });

      await producer.validateRequest(
        "POST",
        "/users?page=1",
        { "content-type": "application/json" },
        { name: "test" },
      );

      expect(capturedInteraction).not.toBeNull();
      expect(capturedInteraction!.method).toBe("POST");
      expect(capturedInteraction!.path).toBe("/users?page=1");
      expect(capturedInteraction!.body).toEqual({ name: "test" });
    });
  });

  describe("validateResponse", () => {
    it("should validate a valid response", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
      });

      const result = await producer.validateResponse(
        "GET",
        "/users/123",
        {},
        null,
        200,
        { "content-type": "application/json" },
        { id: 123, name: "test" },
      );

      expect(result.valid).toBe(true);
      expect(result.type).toBe("response");
    });

    it("should return invalid for invalid response", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createInvalidValidator(["response body type mismatch"]),
      });

      const result = await producer.validateResponse(
        "GET",
        "/users/123",
        {},
        null,
        200,
        {},
        { id: "not-a-number" },
      );

      expect(result.valid).toBe(false);
      expect(result.errors).toContain("response body type mismatch");
      expect(result.type).toBe("response");
    });

    it("should skip validation when validateResponse is false", async () => {
      const validateFn = jest.fn();
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(validateFn),
        validateResponse: false,
      });

      const result = await producer.validateResponse(
        "GET",
        "/users/123",
        {},
        null,
        200,
        {},
        {},
      );

      expect(result.valid).toBe(true);
      expect(validateFn).not.toHaveBeenCalled();
    });
  });

  describe("handleRequestFailure", () => {
    it("should reject request in strict mode", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "strict",
      });

      const req = createMockRequest("POST", "/users");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["test error"],
      };

      const shouldContinue = await producer.handleRequestFailure(
        result,
        req,
        res,
      );

      expect(shouldContinue).toBe(false);
      expect(res.statusCode).toBe(400);
      expect(res.body).toEqual({
        error: "Request validation failed",
        details: ["test error"],
      });
    });

    it("should continue in warn mode", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "warn",
      });

      const req = createMockRequest("POST", "/users");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["test error"],
      };

      // Suppress console.warn
      const warnSpy = jest.spyOn(console, "warn").mockImplementation();

      const shouldContinue = await producer.handleRequestFailure(
        result,
        req,
        res,
      );

      expect(shouldContinue).toBe(true);
      expect(res.body).toBeNull(); // Response not modified
      expect(warnSpy).toHaveBeenCalled();

      warnSpy.mockRestore();
    });

    it("should continue in shadow mode", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "shadow",
      });

      const req = createMockRequest("POST", "/users");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["test error"],
      };

      const shouldContinue = await producer.handleRequestFailure(
        result,
        req,
        res,
      );

      expect(shouldContinue).toBe(true);
      expect(res.body).toBeNull(); // Response not modified
    });

    it("should call custom onRequestFailure handler", async () => {
      const customHandler = jest.fn().mockReturnValue({ custom: "response" });

      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "strict",
        onRequestFailure: customHandler,
      });

      const req = createMockRequest("POST", "/users");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["test error"],
      };

      const shouldContinue = await producer.handleRequestFailure(
        result,
        req,
        res,
      );

      expect(customHandler).toHaveBeenCalledWith(result, req);
      expect(shouldContinue).toBe(false);
    });
  });

  describe("handleResponseFailure", () => {
    it("should log in strict mode", async () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "strict",
      });

      const req = createMockRequest("GET", "/users/123");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["response error"],
      };

      const warnSpy = jest.spyOn(console, "warn").mockImplementation();

      await producer.handleResponseFailure(result, req, res);

      expect(warnSpy).toHaveBeenCalled();
      warnSpy.mockRestore();
    });

    it("should call custom onResponseFailure handler", async () => {
      const customHandler = jest.fn();

      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        mode: "strict",
        onResponseFailure: customHandler,
      });

      const req = createMockRequest("GET", "/users/123");
      const res = createMockResponse();
      const result: ProducerValidationResult = {
        valid: false,
        errors: ["response error"],
      };

      await producer.handleResponseFailure(result, req, res);

      expect(customHandler).toHaveBeenCalledWith(result, req, res);
    });
  });

  describe("shouldValidatePath", () => {
    it("should validate all paths by default", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
      });

      expect(producer.shouldValidatePath("/api/users")).toBe(true);
      expect(producer.shouldValidatePath("/health")).toBe(true);
      expect(producer.shouldValidatePath("/any/path")).toBe(true);
    });

    it("should exclude paths matching excludePaths", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        excludePaths: ["/health", "/metrics"],
      });

      expect(producer.shouldValidatePath("/api/users")).toBe(true);
      expect(producer.shouldValidatePath("/health")).toBe(false);
      expect(producer.shouldValidatePath("/metrics")).toBe(false);
    });

    it("should only include paths matching includePaths", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        includePaths: ["/api/"],
      });

      expect(producer.shouldValidatePath("/api/users")).toBe(true);
      expect(producer.shouldValidatePath("/health")).toBe(false);
    });

    it("should give precedence to excludePaths", () => {
      const producer = new Producer({
        schemaId: "test-schema",
        validator: createMockValidator(),
        includePaths: ["/api/"],
        excludePaths: ["/api/internal"],
      });

      expect(producer.shouldValidatePath("/api/users")).toBe(true);
      expect(producer.shouldValidatePath("/api/internal/debug")).toBe(false);
    });
  });
});

describe("Path filtering utilities", () => {
  describe("matchesPathFilter", () => {
    it("should match string patterns (substring)", () => {
      expect(matchesPathFilter("/api/users", "/api/")).toBe(true);
      expect(matchesPathFilter("/api/users", "/users")).toBe(true);
      expect(matchesPathFilter("/api/users", "/other")).toBe(false);
    });

    it("should match regex patterns", () => {
      expect(matchesPathFilter("/api/users/123", /^\/api\//)).toBe(true);
      expect(matchesPathFilter("/health", /^\/api\//)).toBe(false);
      expect(matchesPathFilter("/users/123", /\/users\/\d+/)).toBe(true);
    });
  });

  describe("shouldValidatePath", () => {
    it("should validate all paths when no filters", () => {
      expect(shouldValidatePath("/any/path")).toBe(true);
      expect(shouldValidatePath("/another/path")).toBe(true);
    });

    it("should apply excludePaths first", () => {
      expect(shouldValidatePath("/health", [], ["/health"])).toBe(false);
      expect(shouldValidatePath("/api/users", [], ["/health"])).toBe(true);
    });

    it("should apply includePaths after excludePaths", () => {
      expect(shouldValidatePath("/api/users", ["/api/"], [])).toBe(true);
      expect(shouldValidatePath("/health", ["/api/"], [])).toBe(false);
    });
  });
});

describe("Metrics", () => {
  beforeEach(() => {
    resetMetrics();
  });

  describe("recordValidationMetrics", () => {
    it("should record request validations", () => {
      recordValidationMetrics("request", { valid: true });
      recordValidationMetrics("request", { valid: false, errors: ["error"] });

      const metrics = getMetrics();
      expect(metrics.requestValidations).toBe(2);
      expect(metrics.requestValidationsPassed).toBe(1);
      expect(metrics.requestValidationsFailed).toBe(1);
    });

    it("should record response validations", () => {
      recordValidationMetrics("response", { valid: true });
      recordValidationMetrics("response", { valid: true });
      recordValidationMetrics("response", { valid: false, errors: ["error"] });

      const metrics = getMetrics();
      expect(metrics.responseValidations).toBe(3);
      expect(metrics.responseValidationsPassed).toBe(2);
      expect(metrics.responseValidationsFailed).toBe(1);
    });
  });

  describe("recordRejection", () => {
    it("should record rejections", () => {
      recordRejection();
      recordRejection();

      const metrics = getMetrics();
      expect(metrics.requestsRejected).toBe(2);
    });
  });

  describe("resetMetrics", () => {
    it("should reset all metrics to zero", () => {
      recordValidationMetrics("request", { valid: true });
      recordValidationMetrics("response", { valid: false, errors: [] });
      recordRejection();

      resetMetrics();

      const metrics = getMetrics();
      expect(metrics.requestValidations).toBe(0);
      expect(metrics.responseValidations).toBe(0);
      expect(metrics.requestsRejected).toBe(0);
    });
  });
});

File: sdks/java/src/test/java/com/cvt/sdk/producer/ProducerTest.java
package com.cvt.sdk.producer;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ProducerTest {

    /**
     * Mock validator that returns configurable results.
     */
    static class MockValidator implements Validator {
        private final boolean valid;
        private final List<String> errors;
        private int callCount = 0;
        private String lastSchemaId;
        private String lastMethod;
        private String lastPath;

        MockValidator(boolean valid, List<String> errors) {
            this.valid = valid;
            this.errors = errors;
        }

        MockValidator(boolean valid) {
            this(valid, Collections.emptyList());
        }

        @Override
        public ProducerValidationResult validate(
                String schemaId,
                String method,
                String path,
                Map<String, String> headers,
                String body,
                int statusCode,
                Map<String, String> responseHeaders,
                String responseBody) {
            callCount++;
            lastSchemaId = schemaId;
            lastMethod = method;
            lastPath = path;
            return new ProducerValidationResult(valid, errors, "");
        }

        int getCallCount() {
            return callCount;
        }

        String getLastSchemaId() {
            return lastSchemaId;
        }

        String getLastMethod() {
            return lastMethod;
        }

        String getLastPath() {
            return lastPath;
        }
    }

    @Nested
    @DisplayName("ProducerConfig Builder Tests")
    class ProducerConfigTest {

        @Test
        @DisplayName("should create config with valid settings")
        void createValidConfig() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.STRICT)
                    .build();

            assertNotNull(config);
            assertEquals("test-schema", config.getSchemaId());
            assertEquals(ValidationMode.STRICT, config.getMode());
            assertTrue(config.isValidateRequest());
            assertTrue(config.isValidateResponse());
        }

        @Test
        @DisplayName("should throw exception when schemaId is missing")
        void throwWhenSchemaIdMissing() {
            assertThrows(IllegalArgumentException.class, () -> {
                ProducerConfig.builder()
                        .validator(new MockValidator(true))
                        .build();
            });
        }

        @Test
        @DisplayName("should throw exception when schemaId is empty")
        void throwWhenSchemaIdEmpty() {
            assertThrows(IllegalArgumentException.class, () -> {
                ProducerConfig.builder()
                        .schemaId("")
                        .validator(new MockValidator(true))
                        .build();
            });
        }

        @Test
        @DisplayName("should throw exception when validator is missing")
        void throwWhenValidatorMissing() {
            assertThrows(IllegalArgumentException.class, () -> {
                ProducerConfig.builder()
                        .schemaId("test-schema")
                        .build();
            });
        }

        @Test
        @DisplayName("should apply default values")
        void applyDefaultValues() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .build();

            assertEquals(ValidationMode.STRICT, config.getMode());
            assertTrue(config.isValidateRequest());
            assertTrue(config.isValidateResponse());
            assertEquals("CVT", config.getLogPrefix());
        }

        @Test
        @DisplayName("should use custom mode")
        void useCustomMode() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.WARN)
                    .build();

            assertEquals(ValidationMode.WARN, config.getMode());
        }
    }

    @Nested
    @DisplayName("Path Filtering Tests")
    class PathFilteringTest {

        @Test
        @DisplayName("should validate all paths by default")
        void validateAllPathsByDefault() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .build();

            assertTrue(config.shouldValidatePath("/api/users"));
            assertTrue(config.shouldValidatePath("/health"));
            assertTrue(config.shouldValidatePath("/any/path"));
        }

        @Test
        @DisplayName("should exclude paths matching excludePath patterns")
        void excludeMatchingPaths() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .excludePath("/health")
                    .excludePath("/metrics")
                    .build();

            assertTrue(config.shouldValidatePath("/api/users"));
            assertFalse(config.shouldValidatePath("/health"));
            assertFalse(config.shouldValidatePath("/metrics"));
        }

        @Test
        @DisplayName("should only validate paths matching includePath patterns")
        void includeMatchingPathsOnly() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .includePath("/api/.*")
                    .build();

            assertTrue(config.shouldValidatePath("/api/users"));
            assertFalse(config.shouldValidatePath("/health"));
        }

        @Test
        @DisplayName("should give precedence to exclude over include")
        void excludeTakesPrecedence() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .includePath("/api/.*")
                    .excludePath("/api/internal.*")
                    .build();

            assertTrue(config.shouldValidatePath("/api/users"));
            assertFalse(config.shouldValidatePath("/api/internal/debug"));
        }
    }

    @Nested
    @DisplayName("Request Validation Tests")
    class RequestValidationTest {

        @Test
        @DisplayName("should validate a valid request")
        void validateValidRequest() {
            MockValidator validator = new MockValidator(true);
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .build();

            Producer producer = new Producer(config);
            Map<String, String> headers = new HashMap<>();
            headers.put("content-type", "application/json");

            ProducerValidationResult result = producer.validateRequest(
                    "POST", "/users", headers, "{\"name\":\"test\"}");

            assertTrue(result.isValid());
            assertEquals("request", result.getType());
            assertEquals(1, validator.getCallCount());
        }

        @Test
        @DisplayName("should return invalid for invalid request")
        void returnInvalidForInvalidRequest() {
            MockValidator validator = new MockValidator(false,
                    Arrays.asList("missing required field: email"));
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .build();

            Producer producer = new Producer(config);

            ProducerValidationResult result = producer.validateRequest(
                    "POST", "/users", null, "{\"name\":\"test\"}");

            assertFalse(result.isValid());
            assertTrue(result.getErrors().contains("missing required field: email"));
        }

        @Test
        @DisplayName("should skip validation when validateRequest is false")
        void skipValidationWhenDisabled() {
            MockValidator validator = new MockValidator(false);
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .validateRequest(false)
                    .build();

            Producer producer = new Producer(config);

            ProducerValidationResult result = producer.validateRequest(
                    "POST", "/users", null, "{}");

            assertTrue(result.isValid());
            assertEquals(0, validator.getCallCount());
        }

        @Test
        @DisplayName("should pass correct data to validator")
        void passCorrectDataToValidator() {
            MockValidator validator = new MockValidator(true);
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("my-api")
                    .validator(validator)
                    .build();

            Producer producer = new Producer(config);

            producer.validateRequest("POST", "/users?page=1", null, null);

            assertEquals("my-api", validator.getLastSchemaId());
            assertEquals("POST", validator.getLastMethod());
            assertEquals("/users?page=1", validator.getLastPath());
        }
    }

    @Nested
    @DisplayName("Response Validation Tests")
    class ResponseValidationTest {

        @Test
        @DisplayName("should validate a valid response")
        void validateValidResponse() {
            MockValidator validator = new MockValidator(true);
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .build();

            Producer producer = new Producer(config);

            ProducerValidationResult result = producer.validateResponse(
                    "GET", "/users/123", null, null,
                    200, null, "{\"id\":123}");

            assertTrue(result.isValid());
            assertEquals("response", result.getType());
        }

        @Test
        @DisplayName("should return invalid for invalid response")
        void returnInvalidForInvalidResponse() {
            MockValidator validator = new MockValidator(false,
                    Arrays.asList("response body type mismatch"));
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .build();

            Producer producer = new Producer(config);

            ProducerValidationResult result = producer.validateResponse(
                    "GET", "/users/123", null, null,
                    200, null, "{\"id\":\"not-a-number\"}");

            assertFalse(result.isValid());
            assertTrue(result.getErrors().contains("response body type mismatch"));
        }

        @Test
        @DisplayName("should skip validation when validateResponse is false")
        void skipValidationWhenDisabled() {
            MockValidator validator = new MockValidator(false);
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(validator)
                    .validateResponse(false)
                    .build();

            Producer producer = new Producer(config);

            ProducerValidationResult result = producer.validateResponse(
                    "GET", "/users/123", null, null,
                    200, null, "{}");

            assertTrue(result.isValid());
            assertEquals(0, validator.getCallCount());
        }
    }

    @Nested
    @DisplayName("Request Failure Handling Tests")
    class RequestFailureHandlingTest {

        private Object mockRequest = new Object();

        @Test
        @DisplayName("should reject request in strict mode")
        void rejectInStrictMode() {
            Producer.resetMetrics();
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.STRICT)
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("test error"), "request");

            Object[] handleResult = producer.handleRequestFailure(result, mockRequest);

            assertFalse((Boolean) handleResult[0]); // Should not continue
            assertNull(handleResult[1]); // No custom response

            Producer.Metrics metrics = Producer.getMetrics();
            assertEquals(1, metrics.getRequestsRejected());
        }

        @Test
        @DisplayName("should continue in warn mode")
        void continueInWarnMode() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.WARN)
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("test error"), "request");

            Object[] handleResult = producer.handleRequestFailure(result, mockRequest);

            assertTrue((Boolean) handleResult[0]); // Should continue
        }

        @Test
        @DisplayName("should continue in shadow mode")
        void continueInShadowMode() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.SHADOW)
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("test error"), "request");

            Object[] handleResult = producer.handleRequestFailure(result, mockRequest);

            assertTrue((Boolean) handleResult[0]); // Should continue
        }

        @Test
        @DisplayName("should call custom onRequestFailure handler")
        void callCustomHandler() {
            AtomicBoolean handlerCalled = new AtomicBoolean(false);
            AtomicReference<ProducerValidationResult> capturedResult = new AtomicReference<>();

            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.STRICT)
                    .onRequestFailure((res, req) -> {
                        handlerCalled.set(true);
                        capturedResult.set(res);
                        return Map.of("custom", "response");
                    })
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("test error"), "request");

            Object[] handleResult = producer.handleRequestFailure(result, mockRequest);

            assertTrue(handlerCalled.get());
            assertNotNull(capturedResult.get());
            assertFalse((Boolean) handleResult[0]); // Should not continue
            assertNotNull(handleResult[1]); // Custom response
        }
    }

    @Nested
    @DisplayName("Response Failure Handling Tests")
    class ResponseFailureHandlingTest {

        private Object mockRequest = new Object();
        private Object mockResponse = new Object();

        @Test
        @DisplayName("should handle response failure in strict mode")
        void handleInStrictMode() {
            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.STRICT)
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("response error"), "response");

            // Should not throw
            assertDoesNotThrow(() ->
                producer.handleResponseFailure(result, mockRequest, mockResponse));
        }

        @Test
        @DisplayName("should call custom onResponseFailure handler")
        void callCustomHandler() {
            AtomicBoolean handlerCalled = new AtomicBoolean(false);

            ProducerConfig config = ProducerConfig.builder()
                    .schemaId("test-schema")
                    .validator(new MockValidator(true))
                    .mode(ValidationMode.STRICT)
                    .onResponseFailure(res -> handlerCalled.set(true))
                    .build();

            Producer producer = new Producer(config);
            ProducerValidationResult result = new ProducerValidationResult(
                    false, Arrays.asList("response error"), "response");

            producer.handleResponseFailure(result, mockRequest, mockResponse);

            assertTrue(handlerCalled.get());
        }
    }

    @Nested
    @DisplayName("Metrics Tests")
    class MetricsTest {

        @BeforeEach
        void resetMetrics() {
            Producer.resetMetrics();
        }

        @Test
        @DisplayName("should record request validations")
        void recordRequestValidations() {
            ProducerValidationResult passResult = ProducerValidationResult.success("request");
            ProducerValidationResult failResult = new ProducerValidationResult(
                    false, Arrays.asList("error"), "request");

            Producer.recordValidationMetrics("request", passResult);
            Producer.recordValidationMetrics("request", failResult);

            Producer.Metrics metrics = Producer.getMetrics();
            assertEquals(2, metrics.getRequestValidations());
            assertEquals(1, metrics.getRequestValidationsPassed());
            assertEquals(1, metrics.getRequestValidationsFailed());
        }

        @Test
        @DisplayName("should record response validations")
        void recordResponseValidations() {
            ProducerValidationResult passResult = ProducerValidationResult.success("response");
            ProducerValidationResult failResult = new ProducerValidationResult(
                    false, Arrays.asList("error"), "response");

            Producer.recordValidationMetrics("response", passResult);
            Producer.recordValidationMetrics("response", passResult);
            Producer.recordValidationMetrics("response", failResult);

            Producer.Metrics metrics = Producer.getMetrics();
            assertEquals(3, metrics.getResponseValidations());
            assertEquals(2, metrics.getResponseValidationsPassed());
            assertEquals(1, metrics.getResponseValidationsFailed());
        }

        @Test
        @DisplayName("should record rejections")
        void recordRejections() {
            Producer.recordRejection();
            Producer.recordRejection();

            Producer.Metrics metrics = Producer.getMetrics();
            assertEquals(2, metrics.getRequestsRejected());
        }

        @Test
        @DisplayName("should reset all metrics")
        void resetAllMetrics() {
            Producer.recordValidationMetrics("request",
                    ProducerValidationResult.success("request"));
            Producer.recordValidationMetrics("response",
                    ProducerValidationResult.success("response"));
            Producer.recordRejection();

            Producer.resetMetrics();

            Producer.Metrics metrics = Producer.getMetrics();
            assertEquals(0, metrics.getRequestValidations());
            assertEquals(0, metrics.getResponseValidations());
            assertEquals(0, metrics.getRequestsRejected());
        }
    }

    @Nested
    @DisplayName("ProducerValidationResult Tests")
    class ProducerValidationResultTest {

        @Test
        @DisplayName("should create success result")
        void createSuccessResult() {
            ProducerValidationResult result = ProducerValidationResult.success("request");

            assertTrue(result.isValid());
            assertTrue(result.getErrors().isEmpty());
            assertEquals("request", result.getType());
        }

        @Test
        @DisplayName("should create failure result")
        void createFailureResult() {
            List<String> errors = Arrays.asList("error1", "error2");
            ProducerValidationResult result = ProducerValidationResult.failure(errors, "response");

            assertFalse(result.isValid());
            assertEquals(2, result.getErrors().size());
            assertEquals("response", result.getType());
        }
    }
}
